featurebyte/query_graph/sql/aggregator/range_join.py

```diff
@@ -0,0 +1,232 @@
+        """
+        Get expressions for the qualified columns
+
+        Returns
+        -------
+        List[Expression]
+        """
+        return [self.get_qualified_column(key) for key in self.columns]
+
+    class Config:
+        """
+        Allow Expression field
+        """
+
+        arbitrary_types_allowed = True
+        extra = "forbid"
+
+
+class LeftTable(BaseTable):
+    """
+    Representation of the left table (typically the request table) in a range join
+    """
+
+    join_keys: list[str]
+    range_start: str
+    range_end: str
+    columns: list[str]
+
+    @property
+    def qualified_range_start(self) -> Expression:
+        """
+        Get an expression for the qualified range start column
+
+        Returns
+        -------
+        Expression
+        """
+        return self.get_qualified_column(self.range_start)
+
+    @property
+    def qualified_range_end(self) -> Expression:
+        """
+        Get an expression for the qualified range end column
+
+        Returns
+        -------
+        Expression
+        """
+        return self.get_qualified_column(self.range_end)
+
+
+class RightTable(BaseTable):
+    """
+    Representation of the right table (e.g. a tile table) in a range join
+    """
+
+    range_column: str
+
+    @property
+    def qualified_range_column(self) -> Expression:
+        """
+        Get an expression for the range column
+
+        Returns
+        -------
+        Expression
+        """
+        return self.get_qualified_column(self.range_column)
+
+
+def range_join_tables(
+    left_table: LeftTable,
+    right_table: RightTable,
+    window_size: int,
+) -> Select:
+    """
+    Join two tables with range join.
+
+    Each row in the left table defines a range start and range end. For each row in the left table,
+    we want to join it with rows in the right table with range column is between range start and
+    range end.
+
+    Parameters
+    ----------
+    left_table: LeftTable
+        Left table in the range join, e.g. the request table
+    right_table: RightTable
+        Right table in the range join, e.g. the tile table
+    window_size: int
+        Window size of the range join
+
+    Returns
+    -------
+    Select
+    """
+
+    # Ultimate condition:
+    # RIGHT.RANGE_COLUMN >= LEFT.RANGE_START AND RIGHT.RANGE_COLUMN < LEFT.RANGE_END
+    range_join_where_conditions = [
+        expressions.GTE(
+            this=right_table.qualified_range_column,
+            expression=left_table.qualified_range_start,
+        ),
+        expressions.LT(
+            this=right_table.qualified_range_column,
+            expression=left_table.qualified_range_end,
+        ),
+    ]
+
+    # Narrow down matches using these conditions before filtering by range_join_where_conditions.
+    # (
+    #   FLOOR(LEFT.RANGE_END / WINDOW) = FLOOR(RIGHT.RANGE_COLUMN / WINDOW)
+    #   AND
+    #   FLOOR(LEFT.RANGE_END / WINDOW) - 1 = FLOOR(RIGHT.RANGE_COLUMN / WINDOW)
+    # )
```

Comment:
Just to double check whether this is correct? If i understood it correctly, this will always be false. 🤔

Reply:
Since union all is used later, maybe this is more like `OR` rather than `AND`.

Reply:
You are right, it should be `OR`. Will correct the comment.